src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -486,19 +475,57 @@ protected void startSessionTracker() {
         ((SessionTrackerImpl)sessionTracker).start();
     }
 
+    /**
+     * Sets the state of ZooKeeper server. After changing the state, this will
+     * notify the server state changes to registered shutdown handler, if any.
+     * <p>
+     * Following are the server state transitions:
+     * <li>During startup the server will be in INITIAL state.</li>
+     * <li>After the successful startup the server will set the state to
+     * RUNNING.</li>
+     * <li>Server will enter into ERROR state if it hits an internal error.
+     * {@link ZooKeeperServerListenerImpl} will notify critical resource error
+     * events.</li>
+     * <li>During shutdown the server will set the state to SHUTDOWN, which
+     * represents that the server is not running.</li>
+     *
+     * @param state new server state.
+     */
+    void setState(State state) {
+        this.state = state;
+        // Notify server state changes to the registered shutdown handler, if any.
+        if (zkShutdownHandler != null) {
```

Comment:
I agree to add an assert. But in case of quorum, 'shutdownHandler' will be null, so it will make noise in quorum. So I'm planning to override #setState() api in subclassess -QuorumZooKeeperServer, ReadOnlyZooKeeperServer.

    @Override
    protected void setState(State state) {
        this.state = state;
    }

For that, some additional changes are required and those are:
(1) Make ZooKeeperServer#setState() api visibility to 'protected'. Now its default, package visibility.
(2) Make ZooKeeperServer's "enum State" visibility to 'protected'

OK?